libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp

```diff
@@ -0,0 +1,615 @@
+      !result) {
+    return result.error();
+  }
+
+  /*
+   * To keep track of communities for nodes in the original graph.
+   * Community will be set to -1 for isolated nodes
+   */
+  LargeArray clusters_orig;
+  clusters_orig.allocateBlocked(pfg->num_nodes());
+
+  LouvainClusteringImplementation<EdgeWeightType> impl{};
+  if (auto r = impl.LouvainClustering(
+          pfg, edge_weight_property_name, clusters_orig, plan);
+      !r) {
+    return r.error();
+  }
+
+  // Remove all the existing node/edge properties
+  if (auto r = pfg->RemoveAllNodeProperties(); !r) {
+    return r.error();
+  }
+
+  if (auto r = ConstructNodeProperties<std::tuple<LCPreviousCommunityId>>(
+          pfg, {output_property_name});
+      !r) {
+    return r.error();
+  }
+
+  auto graph_result = katana::
+      TypedPropertyGraph<std::tuple<LCPreviousCommunityId>, std::tuple<>>::Make(
+          pfg, {output_property_name}, {});
+  if (!graph_result) {
+    return graph_result.error();
+  }
+  auto graph = graph_result.value();
+
+  katana::do_all(
+      katana::iterate(graph),
+      [&](uint32_t i) {
+        graph.GetData<LCPreviousCommunityId>(i) = clusters_orig[i];
+      },
+      katana::loopname("Add clusterIds"), katana::no_stats());
+
+  return katana::ResultSuccess();
+}
+
+}  // anonymous namespace
+
+katana::Result<void>
+katana::analytics::LouvainClustering(
+    katana::PropertyGraph* pg, const std::string& edge_weight_property_name,
+    const std::string& output_property_name, LouvainClusteringPlan plan) {
+  switch (pg->GetEdgeProperty(edge_weight_property_name)->type()->id()) {
+  case arrow::UInt32Type::type_id:
+    return LouvainClusteringWithWrap<uint32_t>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  case arrow::Int32Type::type_id:
+    return LouvainClusteringWithWrap<int32_t>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  case arrow::UInt64Type::type_id:
+    return LouvainClusteringWithWrap<uint64_t>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  case arrow::Int64Type::type_id:
+    return LouvainClusteringWithWrap<int64_t>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  case arrow::FloatType::type_id:
+    return LouvainClusteringWithWrap<float>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  case arrow::DoubleType::type_id:
+    return LouvainClusteringWithWrap<double>(
+        pg, edge_weight_property_name, output_property_name, plan);
+  default:
+    return katana::ErrorCode::TypeError;
+  }
+}
+
+/// \cond DO_NOT_DOCUMENT
+katana::Result<void>
+katana::analytics::LouvainClusteringAssertValid(
+    [[maybe_unused]] katana::PropertyGraph* pg,
+    [[maybe_unused]] const std::string& edge_weight_property_name,
+    [[maybe_unused]] const std::string& property_name) {
+  // TODO(gill): This should have real checks. amp has no idea what to check.
```

Comment:
It's the than amp has no idea what to check, but in this case gill doesn't either. ;-)

Reply:
:P lol